### PR TITLE
`async fn` lifetime elision tests

### PR DESCRIPTION
**Pull request by Centril**
Add `async fn` version of the tests in https://github.com/rust-lang/rust/pull/61207 per the first checkbox in https://github.com/rust-lang/rust/issues/62121#issuecomment-506884048.
Works towards resolving blockers in https://github.com/rust-lang/rust/pull/63209.

r? @nikomatsakis 
cc @cramertj

### CODE REVIEW

File: src/test/ui/self/elision/README.md [hunk 1]
```diff
@@ -42,3 +42,34 @@ In each case, we test the following patterns:
...
+- β ⟹ cannot infer an appropriate lifetime
+- γ ⟹ missing lifetime specifier
+
+| `async` file | Pass? | Conforms to `fn`? | How does it diverge? <br/> `fn` ⟶ `async fn` |
+| --- | --- | --- | --- |
+| `self-async.rs` | ✓ | ✓ | N/A |
+| `struct-async.rs`| ✓ | ✓ | N/A |
+| `alias-async.rs`| ✓ | ✓ | N/A |
+| `assoc-async.rs`| ✓ | ✓ | N/A |
+| `ref-self-async.rs` | X | X | α ⟶ β + γ |
```
Comment (Centril) [inline src/test/ui/self/elision/README.md]: N.B. these cases where things diverge.

------------------

While writing these tests I also hit:
```rust
#![feature(async_await)]
#![feature(nll)]

struct A;

impl A {
    async fn foo(&self, f: &u32) -> &A {
        self
    }
}
```
which passes if you remove `async` but with it you get:
```rust

error[E0106]: missing lifetime specifier
 --> src/lib.rs:7:37
  |
7 |     async fn foo(&self, f: &u32) -> &A {
  |                                     ^
  |
  = note: return-position elided lifetimes require exactly one input-position elided lifetime, found multiple.
```

Comment (nikomatsakis) [reply src/test/ui/self/elision/README.md]: Yeah, this seems like a bug, but I see you filed one already (#63388)

**Review by nikomatsakis — APPROVED** [review]
This looks good to me. We should probably land this PR as it documents current state.

**Pull request merged**